superbench/benchmarks/micro_benchmarks/gpu_copy_performance/gpu_copy.cu

```diff
@@ -746,6 +780,266 @@ void SetSubBenchArgsForDToD(int src_gpu_id, int dst_gpu_id, bool is_read, bool i
     }
 }
 
+// dst_rank: < 0 for all ranks, else for specified rank
+__global__ void SMOneToAllCopyKernel(ulong2 **dst_buffers, ulong2 *src_buffer, uint64_t msg_size, int src_rank,
+                                     int dst_rank, int gpu_count) {
+    uint64_t size_per_dst_rank = msg_size / gpu_count;
+    uint64_t num_blocks_per_dst_rank = gridDim.x / gpu_count;
+    if (dst_rank >= 0 && dst_rank != blockIdx.x / num_blocks_per_dst_rank) {
+        return;
+    } else {
+        dst_rank = blockIdx.x / num_blocks_per_dst_rank;
+    }
+    ulong2 *dst_buffer = dst_buffers[dst_rank];
+    uint64_t src_rank_elem_offset = uint64_t(src_rank) * size_per_dst_rank / sizeof(ulong2);
+    uint64_t dst_rank_elem_offset = uint64_t(dst_rank) * size_per_dst_rank / sizeof(ulong2);
+    ulong2 val;
+
+    uint64_t size_per_block = msg_size / gridDim.x;
+    uint64_t num_elems_per_block = size_per_block / sizeof(ulong2);
+    for (uint64_t idx = num_elems_per_block * blockIdx.x + threadIdx.x; idx < num_elems_per_block * (blockIdx.x + 1);
+         idx += blockDim.x) {
+        FetchULong2(val, src_buffer + idx);
+        StoreULong2(dst_buffer + idx + src_rank_elem_offset - dst_rank_elem_offset, val);
```

Comment:
Looks like you do the reading and writing simultaneously.

Reply:
It's reading from local memory and writing to remote. Currently only support SM write.